### PR TITLE
Chronos:  Add orca distributed function to TF2Forecaster

### PR DESCRIPTION
**Pull request by liangs6212**
Added TF2Forecaster support for orca distributed  and some related unit test.

### New API
 - to_local: Same as `pytorchforecaster.to_local`, convert distributed model to local model.
 - get_model: Reture a keras model, same as `pytorchforecaster.get_model`.

### Known issues
 - input data: `tf.data.Dataset` cannot be entered directly, this issue may not be fixed and is caused by `keras`/`tensorflow` itself.
 - save/load: `TF2Estimator.save` and `TF2Estimator.load` not working properly when model is of type subclass. (Will be fixed in other PRs)
 - error message: Need to improve the Error message

### Doc link: https://liangs.readthedocs.io/en/tf2forecaster-distributed/doc/PythonAPI/Chronos/forecasters.html#lstmforecaster

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 1]
```diff
@@ -14,39 +14,60 @@
...
         """
         Fit(Train) the forecaster.
 
+        Example:
+            >>> # Usage when distributed is True and tf.data.Dataset is used as input
+            >>> def train_tf_dataset(config, batch_size):
+            >>>     data = tf.data.Dataset.from_tensor_slice((train_data))
+            >>>     data = ...
+            >>>     return data
+            >>> forecaster.fit(train_tf_dataset, ...)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:53]: we don't need this exception example

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 2]
```diff
@@ -96,20 +133,35 @@ def predict(self, data, batch_size=32):
...
+            if isinstance(data, np.ndarray):
+                data = np_to_xshards(data)
+            if isinstance(data, TSDataset):
+                input_data, _ = data.to_numpy()
+                data = np_to_xshards(input_data)
+            invalidInputError(not isinstance(data, tf.data.Dataset),
+                              "Will support in the future.")
             yhat = self.internal.predict(data, batch_size=batch_size)
+
+            expand_dim = []
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:153]: add comment here (pytorch and tf's different behavior)

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 3]
```diff
@@ -96,20 +133,35 @@ def predict(self, data, batch_size=32):
...
+            yhat = xshard_to_np(yhat, mode="yhat", expand_dim=expand_dim)
         else:
-            yhat = self.internal(data, training=False).numpy()
+            if not self.fitted:
+                invalidInputError(False,
+                                  "You must call fit or restore first before calling predict!")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:158]: leave this outside

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 4]
```diff
@@ -126,16 +178,22 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+                |
+                | 2. A tf.data.Dataset:
                 | A TFDataset instance which contains x and y with same shape as the tuple.
                 | x's shape is (num_samples, lookback, feature_dim),
                 | y's shape is (num_samples, horizon, target_dim).
-                | 3. A bigdl.chronos.data.tsdataset.TSDataset instance
+                | If set distributed to True, the tf.data.Dataset should not be used as
+                | input directly, a function name with tf.data.Dataset as return value
+                | should be input, and the function should contain two parameters
+                | config and batch_size.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:194]: does not support this one.

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 5]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
                           horizon=self.model_config['future_seq_len'])
-            data = data.to_tf_dataset(shuffle=False, batch_size=batch_size)
 
-        if isinstance(data, tuple):
-            input_data, target = data
+        if self.distributed:
+            if isinstance(data, tuple):
+                data = np_to_data_creator(data)
+            if isinstance(data, TSDataset):
+                data = tsdata_to_data_creator(data)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:220]: shuffle=False

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 6]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+                              "tf.data.Dataset is not supported, "
+                              "please replace with numpy.ndarray or TSDataset instance.")
+            return self.internal.evaluate(data, batch_size=batch_size)
         else:
-            input_data = data
-            target = np.asarray(tuple(map(lambda x: x[1], data.as_numpy_iterator())))
-        yhat = self.internal.predict(input_data, batch_size=batch_size)
+            if not self.fitted:
+                invalidInputError(False,
+                                  "You must call fit or restore first before calling evaluate!")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:228]: move this outside the if else logic

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 7]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+        """
+        Transform a distributed forecaster to a local (non-distributed) one.
+
+        you need to call .to_local() and transform the forecaster to a non-
+        distributed one.
+
+        :return: a forecaster instance.
+        """
+        if not self.distributed:
+            invalidInputError(False, "The forecaster has become local.")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:249]: warning

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 8]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+        model = self.internal.get_model()
+        self.internal.shutdown()
+        self.internal = model
+
+        self.distributed = False
+        self.fitted = True
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:255]: `self.fitted = True` no need here

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 9]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+        return self
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:256]: no need to return self

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 10]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+        Returns the learned Keras model.
 
-        aggregate = 'mean' if multioutput == 'uniform_average' else None
-        return Evaluator.evaluate(self.metrics, y_true=target, y_pred=yhat, aggregate=aggregate)
+        :return: a keras model instance
+        """
+        if self.distributed:
+            return self.internal.get_model()
+        else:
+            return self.internal.model
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:267]: return self.internal

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 11]
```diff
@@ -150,47 +208,100 @@ def evaluate(self, data, batch_size=32, multioutput="raw_values"):
...
+            # TODO: can't save and load, temporarily use save_weight instead, will fix later.
+            from bigdl.dllib.utils.file_utils import is_local_path
+            if is_local_path(checkpoint_file):
+                self.internal.save_weights(checkpoint_file)
+            else:
+                self.internal.save(checkpoint_file)
+        else:
+            if not self.fitted:
+                invalidInputError(False,
+                                  "You must call fit or restore first before calling save!")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:285]: move outside.

File: python/chronos/src/bigdl/chronos/forecaster/tf/utils.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    if shuffle:
+        data = data.cache().shuffle(tuple_data[0].shape[0]).batch(batch_size)
+    else:
+        data = data.batch(batch_size).cache()
+    return data.prefetch(tf.data.AUTOTUNE)
+
+
+def np_to_xshards(data):
+    from bigdl.orca.data import XShards
+    _shards = XShards.partition(data)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/utils.py:53]: not sure if `repartition(partition_num)` is needed.

Comment (liangs6212) [reply python/chronos/src/bigdl/chronos/forecaster/tf/utils.py:49]: `repartition` does not change the prediction results, so there is no need to add.

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 13]
```diff
@@ -61,28 +75,44 @@ def fit(self, data, epochs=1, batch_size=32):
         if isinstance(data, TSDataset):
             if data.lookback is None:
                 data.roll(lookback=self.model_config['past_seq_len'],
-                          horizon=self.model_config['future_seq_len'])
-            data = data.to_tf_dataset(shuffle=True, batch_size=batch_size)
-        if isinstance(data, tuple):
-            self.internal.fit(x=data[0], y=data[1], epochs=epochs, batch_size=batch_size)
+                          horizon=self.model_config['future_seq_len'],
+                          feature_col=data.roll_feature,
+                          target_col=data.roll_target)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:80]: why would a tsdataset has `roll_feature` and `roll_target` still need rolling?

File: python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py [hunk 14]
```diff
@@ -96,20 +126,36 @@ def predict(self, data, batch_size=32):
...
+                          "You must call fit or restore first before calling predict!")
+
+        if self.distributed:
+            if isinstance(data, np.ndarray):
+                data = np_to_xshards(data)
+            if isinstance(data, TSDataset):
+                input_data, _ = data.to_numpy()
+                data = np_to_xshards(input_data)
+            invalidInputError(not isinstance(data, tf.data.Dataset),
+                              "Will support in the future.")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/tf/base_forecaster.py:146]: prediction on tf.data.Dataset will be supported in future.

File: python/chronos/src/bigdl/chronos/forecaster/tf/utils.py (hunk 12 above)
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/tf/utils.py:50]: yeah, I should believe the prediction results is independent of partition number, but it may affect the performance (I guess)

**Review by TheaperDeng — COMMENTED** [review]
others LGTM

File: python/chronos/src/bigdl/chronos/forecaster/tf/utils.py (hunk 12 above)
Comment (liangs6212) [reply python/chronos/src/bigdl/chronos/forecaster/tf/utils.py:51]: `PytorchForecaster's` warning will be added in other pr.

**Review by TheaperDeng — APPROVED** [review]
LGTM
